### PR TITLE
test: bump detox + implement dynamic scroll in e2e tests

### PR DESCRIPTION
**Pull request by cortisiko**
## **Description**
The purpose of this PR is to accomplish two things:

bump detox version (this hasn't been done in a couple of months)
Implement dynamic scroll in tests

Previously, the use of APIs higher than 30 was hindered by an Android blocker on Detox's end. As mentioned in https://github.com/MetaMask/metamask-mobile/pull/6384 

> There is currently an issue when running detox on android 12 and 13 (API 32/33) which prevents the tests from running. The issue is, the tap() action is treated like a tapAndHold() action. See the open issue in wix/detox https://github.com/wix/Detox/issues/3762

With this new detox version bump, the blocker has been lifted, allowing us to utilize Android 31+. 

Furthermore, how we scroll to elements introduces flakiness. Instead of hardcoding a scroll speed in tests, we will now introduce dynamic scrolling. We will scroll on a view until the element is visible within the viewport. This was done using [whileElement](https://wix.github.io/Detox/docs/api/actions/#whileelementelement) 
 

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

### **After**

Smoke run: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/42936d4e-1df8-40ca-bd88-4f408cdd655a

Regression Run: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/cb83ae29-a0a4-49f5-a2f5-7c6683bf38e3

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 1]
```diff
@@ -11,31 +10,28 @@ import Matchers from '../../../../utils/Matchers';
 import Gestures from '../../../../utils/Gestures';
 
 class SecurityAndPrivacy {
-  get container() {
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:14]: Removed this element since it is no longer necessary for scrolling.

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 2]
```diff
@@ -11,31 +10,28 @@ import Matchers from '../../../../utils/Matchers';
...
-    );
-  }
-
   get changePasswordButton() {
     return Matchers.getElementByID(
       SecurityPrivacyViewSelectorsIDs.CHANGE_PASSWORD_BUTTON,
     );
   }
 
-  get deleteWalletButtonLabel() {
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:26]: This is no longer needed. We will have a conditional statement in the getter for deleteWallet that sets the appropriate element matcher based on platform

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 3]
```diff
@@ -76,66 +72,35 @@ class SecurityAndPrivacy {
...
-    if (device.getPlatform() === 'android') {
-      await Gestures.waitAndTap(this.deleteWalletButtonLabel);
-    } else {
-      await Gestures.waitAndTap(this.deleteWalletButtonID);
-    }
+    await Gestures.waitAndTap(this.deleteWalletButton);
   }
 
   async scrollToChangePasswordView() {
-    // Scroll to the bottom
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:87]: Removed unwanted actions since we now scroll dynamically.

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 4]
```diff
@@ -76,66 +72,35 @@ class SecurityAndPrivacy {
...
-    }
+    await Gestures.scrollToElement(
+      this.changePasswordButton,
+      this.scrollViewIdentifier,
+    );
   }
 
   async scrollToDeleteWalletButton() {
-    // Scroll to the bottom
-    await Gestures.swipe(this.securitySettingsScroll, 'up', 'fast', 0.6);
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:98]: Removed unwanted actions since we now scroll dynamically.

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 5]
```diff
@@ -76,66 +72,35 @@ class SecurityAndPrivacy {
...
-    }
+    await Gestures.scrollToElement(
+      this.deleteWalletButton,
+      this.scrollViewIdentifier,
+    );
   }
 
   async scrollToTurnOnRememberMe() {
-    // Scroll to the bottom
-    if (device.getPlatform() === 'android') {
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:112]: Removed unwanted actions since we now scroll dynamically.

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 6]
```diff
@@ -76,66 +72,35 @@ class SecurityAndPrivacy {
...
-    }
-    //await TestHelpers.swipe(PRIVACY_MODE_SECTION_ID, 'up', 'fast');
+    await Gestures.scrollToElement(
+      this.rememberMeToggle,
+      this.scrollViewIdentifier,
+    );
   }
 
   async scrollToMetaMetrics() {
-    await Gestures.swipe(this.backUpNow, 'up', 'fast', 0.6);
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:122]: Removed unwanted actions since we now scroll dynamically.

File: e2e/specs/permission-systems/permission-system-delete-wallet.spec.js [hunk 7]
```diff
@@ -52,13 +52,14 @@ describe(
 
           //go to settings then security & privacy
           await TabBarComponent.tapSettings();
-          await SettingsView.tapSecurityAndPrivacy();
```
Comment (cortisiko) [inline e2e/specs/permission-systems/permission-system-delete-wallet.spec.js:55]: Restructured these test steps because the test is hanging when it gets to the bottom of the security and privacy view. We can follow up in another PR to investigate and address this.

File: e2e/utils/Gestures.js [hunk 8]
```diff
@@ -154,16 +154,25 @@ class Gestures {
...
-   * @param {number} distance - Distance to scroll
-   * @param {Detox.Direction} direction - Direction of the scroll (up, down, left, right)
-   */
-  static async scrollUpTo(elementID, distance, direction) {
-    const element = await elementID;
+   * @param {Promise<Detox.IndexableNativeElement>} destinationElementID - ID of the element to scroll up to
+   * @param {number} scrollIdentifier - The identifier (by.id) NOT elementID (element(by.id)). Keep this distinction in mind. If you pass in an elementID this method would not work as intended
+   * @param {Detox.Direction} direction - Direction of the scroll (up, down, left, right). The default is down.
+   * @param {number} [scrollAmount=350] - The amount to scroll (default is 350). Optional parameter.   */
+  static async scrollToElement(
```
Comment (cortisiko) [inline e2e/utils/Gestures.js:163]: Here we have a scrollToElement method:

It takes in the _Element_ we wish to scroll to. 
The _identifier_  or better yet, as detox calls it _Matcher_ of the element. Keep in mind.

What is the difference between an element and an identifier


An element() is used to create objects that represent UI elements. When we create an element, we can perform different actions like tap, long press, swipe etc. See Example in our code [here](https://github.com/MetaMask/metamask-mobile/blob/2f426e8e4238c4b736941614988da31a3959338d/e2e/utils/Matchers.js#L14)

A matcher: is the method that specifies the type of selector needed to locate an element. Read up more on Matchers [here](https://wix.github.io/Detox/docs/api/matchers) 

Understanding this distinction is crucial in order to effectively utilize [whileElement](https://wix.github.io/Detox/docs/api/actions/#whileelementelement)

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 9]
```diff
@@ -76,66 +72,35 @@ class SecurityAndPrivacy {
   }
 
   async tapDeleteWalletButton() {
-    if (device.getPlatform() === 'android') {
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:79]: This conditional is now done in the getter

File: e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js [hunk 10]
```diff
@@ -11,31 +10,28 @@ import Matchers from '../../../../utils/Matchers';
...
+      ? Matchers.getElementByID(SECURITY_PRIVACY_DELETE_WALLET_BUTTON)
+      : Matchers.getElementByLabel(SECURITY_PRIVACY_DELETE_WALLET_BUTTON);
   }
 
   get metaMetricsToggle() {
     return Matchers.getElementByID(
       SecurityPrivacyViewSelectorsIDs.METAMETRICS_SWITCH,
     );
   }
+  get scrollViewIdentifier() {
```
Comment (cortisiko) [inline e2e/pages/Drawer/Settings/SecurityAndPrivacy/SecurityAndPrivacyView.js:30]: Instead of getting the scroll view element, we are getting the identifier for the scroll view. This is important for dynamic scrolling.

**Comment by Cal-L** [comment]
What's the difference between the scroll before and dynamic scrolling?

File: .detoxrc.js [hunk 11]
```diff
@@ -8,7 +8,7 @@ module.exports = {
     jest: {
       setupTimeout: 220000,
     },
-    retries: 2,
+    retries: process.env.CI ? 2 : 1,
```
Comment (Cal-L) [inline .detoxrc.js:11]: Is this change needed?

**Review by Cal-L — COMMENTED** [review]
left some comments

**Comment by cortisiko** [comment]
> What's the difference between the scroll before and dynamic scrolling?

The way we scrolled before was non-deterministic. Depending on various screen sizes, whenever we scroll, the element may or may not be visible. This PR introduces a way to scroll by a defined number of pixels until the desired element is visible.

File: .detoxrc.js (hunk 11 above)
Comment (cortisiko) [reply .detoxrc.js:11]: ah good catch, this was unintended.

**Review by SamuelSalas — APPROVED** [review]
LGTM

**Review by Cal-L — APPROVED** [review]
Lgtm

**Review by chrisleewilcox — APPROVED** [review]
LGTM
